### PR TITLE
compress embedded bootloader

### PR DESCRIPTION
**Pull request by TychoVrahe**
Save flash space: `SUMMARY: 63 rows, -38_400 bytes in total (L-38_424 D24).`

### CODE REVIEW

File: core/site_scons/tools.py [hunk 1]
```diff
@@ -120,3 +121,43 @@ def get_defs_for_cmake(defs: list[str | tuple[str, str]]) -> list[str]:
...
+    z = zlib.compressobj(level=9, wbits=-10)
+    return z.compress(data) + z.flush()
+
+
+def embed_binary(obj_program, env, section, target, file):
+    file_path = PROJECT_ROOT / Path(file)
+
+    with open(file_path, "rb") as f:
+        data = f.read()
+        compressed = _compress(data)
```
Comment (matejcik) [inline core/site_scons/tools.py:136]: quick tip:
```suggestion
    compressed = _compress(file_path.read_bytes())
```
(also `write_bytes()` `write_text()` `read_text()`)

File: core/embed/firmware/bl_check.c [hunk 2]
```diff
@@ -127,14 +149,24 @@ void check_and_replace_bootloader(void) {
...
-      (const uint32_t *)&_binary_embed_firmware_bootloader_bin_start;
+      (const uint32_t
+           *)&_binary_embed_firmware_bootloaders_bootloader_zip_start;
   const uint32_t len =
-      (const uint32_t)&_binary_embed_firmware_bootloader_bin_size;
+      (const uint32_t)&_binary_embed_firmware_bootloaders_bootloader_zip_size;
+
+  struct uzlib_uncomp decomp = {0};
+  uint8_t decomp_window[UZLIB_WINDOW_SIZE] = {0};
+  uint8_t decomp_out[IMAGE_HEADER_SIZE] = {0};
```
Comment (cepetr) [inline core/embed/firmware/bl_check.c:159]: decomp_out should be aligned to a 32-bit boundary, as you later cast it to uint32_t*

File: core/embed/firmware/bl_check.c [hunk 3]
```diff
@@ -174,12 +206,27 @@ void check_and_replace_bootloader(void) {
...
-        NULL);
-  }
-  for (int i = len / sizeof(uint32_t); i < 128 * 1024 / sizeof(uint32_t); i++) {
+
+  uint32_t j = 0;
+  uint32_t d_len = 0;
+
+  do {
+    d_len = decomp.dest - decomp_out;
+    for (int i = 0; i < d_len / sizeof(uint32_t); i++) {
```
Comment (cepetr) [inline core/embed/firmware/bl_check.c:215]: If the length of the last decompressed block is not aligned to a 32-bit boundary, the last few bytes are left unwritten

File: core/site_scons/tools.py [hunk 4]
```diff
@@ -120,3 +121,48 @@ def get_defs_for_cmake(defs: list[str | tuple[str, str]]) -> list[str]:
...
+    start_src = (
+        "_binary_build_firmware_" + _in.replace("/", "_").replace(".", "_") + "_start"
+    )
+    end_src = (
+        "_binary_build_firmware_" + _in.replace("/", "_").replace(".", "_") + "_end"
+    )
+    size_src = (
+        "_binary_build_firmware_" + _in.replace("/", "_").replace(".", "_") + "_size"
+    )
+    start_dest = "_binary_" + target_.replace("/", "_").replace(".o", "_zip") + "_start"
```
Comment (cepetr) [inline core/site_scons/tools.py:142]: 'zip' file extension might be confusing as the file is not a real ZIP file. But this is really a nit.

**Review by cepetr — CHANGES_REQUESTED** [review]

File: core/embed/firmware/bl_check.c (hunk 2 above)
Comment (TychoVrahe) [reply core/embed/firmware/bl_check.c]: fixed https://github.com/trezor/trezor-firmware/pull/3304/commits/9143bf167745d860251791ae305470e98e153266

File: core/embed/firmware/bl_check.c (hunk 3 above)
Comment (TychoVrahe) [reply core/embed/firmware/bl_check.c]: fixed https://github.com/trezor/trezor-firmware/pull/3304/commits/9143bf167745d860251791ae305470e98e153266

File: core/embed/firmware/bl_check.c [hunk 5]
```diff
@@ -174,12 +206,27 @@ void check_and_replace_bootloader(void) {
...
+    }
+    decomp.dest = decomp_out;
+    j++;
+  } while (uzlib_uncompress(&decomp) >= 0);
+
+  // zero out the rest of the bootloader area
+  uint32_t start_zero = (j - 1) * IMAGE_HEADER_SIZE + d_len;
+
+  for (uint32_t i = start_zero / sizeof(uint32_t);
+       i < bl_len / sizeof(uint32_t); i++) {
```
[the remark this answers is not in the document]
Comment (TychoVrahe) [reply core/embed/firmware/bl_check.c]: using with a few mods: https://github.com/trezor/trezor-firmware/pull/3304/commits/9143bf167745d860251791ae305470e98e153266

this code suffers similar problem as the original - if the last block is unaligned it writes garbage (residuals from previous decompression block) in the last few bytes, so i added a solution to that.

(bootloader is always aligned to 512 bytes, which is guaranteed by linker script and checked somewhere in the checks above this - so this shouldn't really matter)

File: core/site_scons/tools.py (hunk 4 above)
Comment (TychoVrahe) [reply core/site_scons/tools.py]: changed to 'bin', open to better suggestions https://github.com/trezor/trezor-firmware/commit/030e89abb801560a6223c0dcf6d5969ea6b6dacd

Comment (cepetr) [reply core/site_scons/tools.py]: I looked around but didn't find anything commonly used. I would like to propose using `.bin.deflated` to distinguish it from an uncompressed file.

Comment (TychoVrahe) [reply core/site_scons/tools.py]: `.bin.deflated` seems good enough: https://github.com/trezor/trezor-firmware/commit/86ebbccd993ae23d00096089c15e698c87fb8e23

**Review by cepetr — APPROVED** [review]

File: core/site_scons/tools.py [hunk 6]
```diff
@@ -120,3 +121,54 @@ def get_defs_for_cmake(defs: list[str | tuple[str, str]]) -> list[str]:
...
+    )
+    start_dest = (
+        "_binary_" + target_.replace("/", "_").replace(".o", "_bin_deflated") + "_start"
+    )
+    end_dest = (
+        "_binary_" + target_.replace("/", "_").replace(".o", "_bin_deflated") + "_end"
+    )
+    size_dest = (
+        "_binary_" + target_.replace("/", "_").replace(".o", "_bin_deflated") + "_size"
+    )
```
Comment (matejcik) [inline core/site_scons/tools.py:150]:
```suggestion
    def redefine_sym(name):
        src = (
            "_binary_build_firmware_" + _in.replace("/", "_").replace(".", "_") + "_" + name
        )
        dest = (
            "_binary_" + target_.replace("/", "_").replace(".o", "_bin_deflated") + "_" + name
        )
        return f" --redefine-sym {src}={dest}"
```

File: core/embed/firmware/bl_check.c [hunk 7]
```diff
@@ -174,16 +208,31 @@ void check_and_replace_bootloader(void) {
...
-  for (int i = len / sizeof(uint32_t); i < 128 * 1024 / sizeof(uint32_t); i++) {
-    ensure(flash_area_write_word(&BOOTLOADER_AREA, i * sizeof(uint32_t),
-                                 0x00000000),
-           NULL);
+
+  uint32_t offset = 0;
+
+  do {
+    uint32_t *p = decomp_out;
+    while ((uint32_t)p < (((uint32_t)decomp.dest) & ~3)) {
```
Comment (matejcik) [inline core/embed/firmware/bl_check.c:216]: `(((uint32_t)decomp.dest) & ~3)`

please spin this out into a local variable with a descriptive name

File: core/embed/firmware/bl_check.c [hunk 8]
```diff
@@ -127,14 +151,24 @@ void check_and_replace_bootloader(void) {
...
+      (const uint32_t)&_binary_embed_firmware_bootloaders_bootloader_bin_deflated_size;
+
+  struct uzlib_uncomp decomp = {0};
+  uint8_t decomp_window[UZLIB_WINDOW_SIZE] = {0};
+  uint32_t decomp_out[IMAGE_HEADER_SIZE / sizeof(uint32_t)] = {0};
+
+  uzlib_prepare(&decomp, decomp_window, data, len, decomp_out,
+                sizeof(decomp_out));
+
+  uzlib_uncompress(&decomp);
```
Comment (matejcik) [inline core/embed/firmware/bl_check.c:166]: we should probably `ensure()` that the decompression went well enough to give us the full header

**Review by matejcik — CHANGES_REQUESTED** [review]

File: core/site_scons/tools.py (hunk 6 above)
Comment (TychoVrahe) [reply core/site_scons/tools.py]: fixed https://github.com/trezor/trezor-firmware/pull/3304/commits/4ddf5b306c2cedbbacdc4aad9c8bfe86f9432184

File: core/embed/firmware/bl_check.c (hunk 7 above)
Comment (TychoVrahe) [reply core/embed/firmware/bl_check.c]: fixed https://github.com/trezor/trezor-firmware/pull/3304/commits/4ddf5b306c2cedbbacdc4aad9c8bfe86f9432184

File: core/embed/firmware/bl_check.c (hunk 8 above)
Comment (TychoVrahe) [reply core/embed/firmware/bl_check.c]: fixed https://github.com/trezor/trezor-firmware/pull/3304/commits/4ddf5b306c2cedbbacdc4aad9c8bfe86f9432184

**Review by matejcik — APPROVED** [review]
flake8 is unhappy, otherwise lgtm

**Review by mmilata — APPROVED** [review]
9f145bb8cb15e02b6befd2dfbf05ae866da90e02 ACK

**Pull request merged**